MudAppBar: Add XML Documentation for Public Members

jperson2000 opened the pull request:
## Description

This update adds XML documentation for the `MudAppBar` component for all public members.  Also, a "Usage" example was added with a `MudSlider` for `Elevation` and a `MudSwitch` for `Dense`, so users can play with different values.  

![image](https://github.com/MudBlazor/MudBlazor/assets/18043079/3cb8d726-cc22-4eaf-86ff-bb76f343dc7d)

![image](https://github.com/MudBlazor/MudBlazor/assets/18043079/18b89255-597c-491c-9bea-8e790a2e8ca7)

## How Has This Been Tested?

This was tested by running unit tests and observing the documentation tooltips shown in Visual Studio 2022 when working with a `MudAppBar` component, like this:

![image](https://github.com/MudBlazor/MudBlazor/assets/18043079/abe07ff1-4cf9-4b54-a12c-e61bcd2476a7)

![image](https://github.com/MudBlazor/MudBlazor/assets/18043079/9ca9795c-2ad2-4ce8-a4fd-e71e4eb424a1)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [X] The PR is submitted to the correct branch (`dev`).
- [X] My code follows the code style of this project.
- [ ] I've added relevant tests.

jperson2000 committed:
MudAppBar: Added XML documentation for public and protected members.

jperson2000 committed:
MudAppBar: Added "Elevation" slider to elevation example.  Changed max from 25 to 24  (25 didn't work)

jperson2000 committed:
MudAppBar: Added MudSwitch so users can play with the "Dense" property.

danielchalmers commented on src/MudBlazor.Docs/Pages/Components/AppBar/Examples/AppBarDenseExample.razor:
```diff
@@ -1,7 +1,14 @@
+@code
+{
+    bool dense = true;
+}
```
By default the code is shown so this might take up too much space

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/805ac890-2610-4801-93a3-b2ac2a444761)

especially if multiplied

jperson2000 replied:
What do you think: would the solution here be to back out the `MudSwitch` addition, or to hide the code?  (I had originally added it because I wanted to see the difference of dense versus regular)

jperson2000 replied:
I added a `ShowCode="false"` to the examples to hide the code by default.  Let me know if that looks ok; happy to roll back if needed.

danielchalmers replied:
Maybe Dense and Elevation examples should be combined like [ToggleGroup Usage](https://mudblazor.com/components/togglegroup#usage) or [Badge Playground](https://mudblazor.com/components/badge#playground)?

Would have to defer to @ScarletKuro on this

jperson2000 replied:
Ahh ok, so perhaps I would roll back the Elevation/Dense changes, but add a "Usage" example at the top to demo all of the settings at once.  I like that!   Like giving users the ability to play with controls before diving into more specific details as they scroll down.  I'll wait to hear from @ScarletKuro

henon replied:
I like the usage example. Of course it would look better if the controls to play with would all be in one wrap panel so they take up just a single line in desktop mode and wrap around on small screens.

danielchalmers commented on src/MudBlazor/Components/AppBar/MudAppBar.razor.cs:
```diff
@@ -3,9 +3,21 @@
+        /// <summary>
+        /// Gets the CSS classes to apply to this component.
+        /// </summary>
+        /// <remarks>
+        /// The set of classes returned depends on the values of the <see cref="Dense"/>, <see cref="Fixed"/>, <see cref="Bottom"/>, <see cref="Elevation"/>, and <see cref="Color"/> properties.
+        /// </remarks>
```
I would be concerned about this going stale too quickly and people simply not maintaining it. Self-documenting code is probably the goal for non-public stuff

jperson2000 replied:
Ah gotcha, okay, I'll back those two `protected` docs out.

danielchalmers commented on src/MudBlazor/Components/AppBar/MudAppBar.razor.cs:
```diff
@@ -16,69 +28,93 @@ public partial class MudAppBar : MudComponentBase
-        /// The higher the number, the heavier the drop-shadow. 0 for no shadow.
+        /// Gets or sets the size of the drop shadow.
         /// </summary>
+        /// <remarks>
+        /// Defaults to 4.  A higher number creates a heavier drop shadow.  Use a value of 0 for no shadow.
+        /// </remarks>
```
An idea:

```xml
/// <remarks>
/// Defaults to <c>4</c>.
/// A higher number creates a heavier drop shadow.
/// Use a value of <c>0</c> for no shadow.
/// </remarks>
```

Splitting them up can be easier to read when changing the source (subjective!) and the `<c>` tags show special formatting when viewed.

jperson2000 replied:
I like the `<c>` idea.  One thing to note is that carriage returns in remarks will also show as carriage returns in the API docs pages, so I just want to make sure everyone's okay with the taller rows.  But I totally agree on the readability.  If perhaps the API docs CSS didn't include the carriage returns, then we could have the best of both.

danielchalmers replied:
> carriage returns in remarks will also show as carriage returns in the API docs pages

Oh that's interesting, I had no idea. I would honestly say that should be fixed on the docs side so it matches Visual Studio 🤷‍♀️ - @ScarletKuro ?

I don't think the docs supports a lot of things: `<see>`, `<c>`, `<para>`

danielchalmers replied:
Imagine if property names linked to the source on GitHub as well

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/22ce9c0d-f783-41df-a4c2-a930b3467a5b)

🤩

danielchalmers commented on src/MudBlazor/Components/AppBar/MudAppBar.razor.cs:
```diff
@@ -16,69 +28,93 @@ public partial class MudAppBar : MudComponentBase
+        /// Gets or sets a value indicating whether left and right padding is added to the appbar.
         /// </summary>
+        /// <remarks>
+        /// Defaults to true.
```
Could use [`<c>`](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/xmldoc/recommended-tags#c) if you wanted

jperson2000 replied:
I'll add these and update the `MudAlert` PR to use these, too, to be consistent.

jperson2000 committed:
MudAppBar: Added <c> tags to colorize literals in docs.  Removed docs on protected members

jperson2000 committed:
MudAppBar: Hid the code for two examples by default now that they are taller

jperson2000 committed:
Merge branch 'dev' into feature/mudappbar-xmldocumentation

jperson2000 commented:
This was put back to Draft while we work out the examples.  It sounds like I should back out my changes to the Elevation and Dense examples but add a "Usage" example at the top (like other docs pages) so people can play with options.  I should have that done tomorrow (4/12).

jperson2000 committed:
MudAppBar: Rolled back changes to Elevation/Dense example.  Added a "Usage" example with Elevation slider and switch for Dense.

jperson2000 committed:
Merge branch 'dev' into feature/mudappbar-xmldocumentation

jperson2000 commented:
Ok @ScarletKuro @henon this is ready to review when you get time.  After all the discussion, the changes are:

* Adding a "Usage" example with a `MudSlider` for `Elevation` and a `MudSwitch` for `Dense`, so users can play with options
* Changing the `Elevation` example to use a value of `24` instead of `4`, to demo a non-default value.
* XML documentation added for all public members in the `MudAppBar` class.

henon commented on src/MudBlazor.Docs/Pages/Components/AppBar/AppBarPage.razor:
```diff
@@ -34,7 +45,7 @@
             <SectionHeader Title="Dense">
                 <Description></Description>
             </SectionHeader>
-            <SectionContent DarkenBackground="true" Code="AppBarDenseExample">
+            <SectionContent DarkenBackground="true" Code="AppBarDenseExample" ShowCode="false">
```
If we don't have anything to write here I guess we can do away with this example. It is already covered by the Usage Example.

henon reviewed (changes requested):
The Usage example needs some layout refinement, otherwise very nice.

henon commented:
@jperson2000 I think we should follow the following style on the doc pages of most components: Start with a very basic example that is so short that we can show the source by default. Ideally this example doesn't need a code block. Then follow with a sophisticated usage example like the one you built. For reference see TreeView:

![image](https://github.com/MudBlazor/MudBlazor/assets/44090/56f08ae6-5bee-4a91-abff-0085d4eff79b)

jperson2000 committed:
Merge branch 'MudBlazor:dev' into feature/mudappbar-xmldocumentation

jperson2000 commented:
> @jperson2000 Ideally this example doesn't need a code block. Then follow with a sophisticated usage example like the one you built. 

Thanks for the great feedback, @henon .  Ok, so the latest examples now have "Basic App Bar" at the top:

![image](https://github.com/MudBlazor/MudBlazor/assets/18043079/3cb8d726-cc22-4eaf-86ff-bb76f343dc7d)

... and then a "Usage" example next, with **Elevation** and **Dense** controls in a MudGrid:

![image](https://github.com/MudBlazor/MudBlazor/assets/18043079/18b89255-597c-491c-9bea-8e790a2e8ca7)

Is this what you were looking for?

jperson2000 committed:
MudAppBar: Combined examples and improved Usage example.

henon commented:
Yes, this is great. 

The controls for the usage demo would look a lot better if they line up nicely. I marked where the Elevation text and the switch should be:
![image](https://github.com/MudBlazor/MudBlazor/assets/44090/52903f66-f3ed-4a83-9eaf-557dedd06138)

If the controls have a top margin so that they have the same distance to the appbar like the slider now has that would be best.

henon commented:
the usage example can hide its code by default.

jperson2000 committed:
MudAppBar: Added mt-6 to MudSwitch and added ShowCode="false" for Usage example.

jperson2000 commented:
> the usage example can hide its code by default.

Ok, I've added `ShowCode="false"` to the usage example, and added `mt-6` to the `MudSwitch` to line the Dense control up vertically:

![image](https://github.com/MudBlazor/MudBlazor/assets/18043079/44797e2f-8341-49d8-ad52-e3dbe660a512)

henon commented:
Perfect. Thanks!
